File: library/src/main/java/com/lzx/library/EfficientAdapter.java
package com.lzx.library;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.util.SparseArray;
import android.view.View;
import android.view.ViewGroup;


import com.lzx.library.differ.AsyncListDiffer;

import java.util.ArrayList;
import java.util.List;

public class EfficientAdapter extends RecyclerView.Adapter<BaseViewHolder> {

    private static final int DEFAULT_ITEM_TYPE = 0; //int 的默认值是0
    private Context mContext;
    private Object[] mObjects;
    private List<MultiTypeEntity> mDataList;
    private SparseArray<IViewHolderCreator> typeHolders;
    private boolean isUseDiffUtil = true;
    private AsyncListDiffer<MultiTypeEntity> mHelper;

    public EfficientAdapter(Context context) {
        initAdapter(context, true);
    }

    public EfficientAdapter(Context context, boolean isUseDiffUtil) {
        initAdapter(context, isUseDiffUtil);
    }

    public EfficientAdapter(Context context, Object[] objects) {
        mObjects = objects;
        initAdapter(context, true);
    }

    public EfficientAdapter(Context context, boolean isUseDiffUtil, Object[] objects) {
        mObjects = objects;
        initAdapter(context, isUseDiffUtil);
    }

    public void setObjects(Object[] objects) {
        mObjects = objects;
    }

    private void initAdapter(Context context, boolean isUseDiffUtil) {
        mContext = context;
        this.isUseDiffUtil = isUseDiffUtil;
        typeHolders = new SparseArray<>();
        if (isUseDiffUtil) {
            mHelper = new AsyncListDiffer<>(this, new AsyncListDiffer.ListChangedCallback<MultiTypeEntity>() {
                @Override
                public void onListChanged(List<MultiTypeEntity> currentList) {
                    mDataList = currentList;
                }
            }, new DiffUtil.ItemCallback<MultiTypeEntity>() {
                // 返回两个item是否相同
                // 例如：此处两个item的数据实体是User类，所以以id作为两个item是否相同的依据
                // 即此处返回两个user的id是否相同
                @Override
                public boolean areItemsTheSame(@NonNull MultiTypeEntity oldItem, @NonNull MultiTypeEntity newItem) {
                    return oldItem.areItemsTheSame(newItem);
                }

                // 当areItemsTheSame返回true时，我们还需要判断两个item的内容是否相同
                // 此处以User的age作为两个item内容是否相同的依据
                // 即返回两个user的age是否相同
                @Override
                public boolean areContentsTheSame(@NonNull MultiTypeEntity oldItem, @NonNull MultiTypeEntity newItem) {
                    return oldItem.areContentsTheSame(newItem);
                }
            });
        }
    }

    public <T extends MultiTypeEntity> EfficientAdapter register(int viewType, final HolderInjector<T> injector) {
        typeHolders.put(viewType, new IViewHolderCreator<T>() {
            @Override
            public BaseViewHolder<T> create(ViewGroup parent) {
                return new BaseViewHolder<T>(parent, injector.getLayoutRes()) {

                    @Override
                    protected void setItemView(View itemView) {
                        injector.setItemView(itemView);
                    }

                    @Override
                    protected void onBindPayloads(Context context, List<Object> payloads, int position, Object... objects) {
                        injector.onInjectUpdate(context, (T) payloads.get(0), position, objects);
                    }

                    @Override
                    protected void onBind(Context context, T data, int position, Object... objects) {
                        injector.onInject(context, data, position, objects);
                    }
                };
            }
        });
        return this;
    }

    public <T extends MultiTypeEntity> EfficientAdapter register(final HolderInjector<T> injector) {
        register(DEFAULT_ITEM_TYPE, injector);
        return this;
    }

    public <T> void submitList(@Nullable List<T> list) {
        if (list == null) {
            return;
        }
        if (isUseDiffUtil) {
            this.mHelper.submitList((List<MultiTypeEntity>) list);
        } else {
            mDataList = (List<MultiTypeEntity>) list;
            notifyDataSetChanged();
        }
    }

    public List<? extends MultiTypeEntity> getCurrentList() {
        if (isUseDiffUtil) {
            return new ArrayList<>(mHelper.getCurrentList());
        } else {
            return mDataList;
        }
    }

    public <T extends MultiTypeEntity> void insertedData(int position, T data) {
        if (isUseDiffUtil) {
            List<MultiTypeEntity> newList = new ArrayList<>(mDataList);
            newList.add(position, data);
            submitList(newList);
        } else {
            mDataList.add(position, data);
            notifyItemInserted(position);
        }
    }

    public void removedData(int position) {
        if (isUseDiffUtil) {
            List<MultiTypeEntity> newList = new ArrayList<>(mDataList);
            newList.remove(position);
            submitList(newList);
        } else {
            mDataList.remove(position);
            notifyItemRemoved(position);
        }
    }

    public <T extends MultiTypeEntity> void removedData(T data) {
        if (isUseDiffUtil) {
            List<MultiTypeEntity> newList = new ArrayList<>(mDataList);
            newList.remove(data);
            submitList(newList);
        } else {
            int removePosition = mDataList.indexOf(data);
            mDataList.remove(data);
            notifyItemRemoved(removePosition);
        }
    }

    public <T extends MultiTypeEntity> void updateData(int position, T data) {
        updateData(position, data, false);
    }

    public <T extends MultiTypeEntity> void updateData(int position, T data, boolean isEfficient) {
        if (isUseDiffUtil) {
            List<MultiTypeEntity> newList = new ArrayList<>(mDataList);
            newList.set(position, data);
            mHelper.updateOldData(newList);
        } else {
            mDataList.set(position, data);
        }
        if (isEfficient) {
            notifyItemChanged(position, data);
        } else {
            notifyItemChanged(position);
        }
    }

    public void clear() {
        if (isUseDiffUtil) {
            mHelper.submitList(null);
        } else {
            if (mDataList != null) {
                mDataList.clear();
            }
        }
    }

    private interface IViewHolderCreator<T extends MultiTypeEntity> {
        BaseViewHolder<T> create(ViewGroup parent);
    }

    @NonNull
    @Override
    public BaseViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        IViewHolderCreator creator = typeHolders.get(viewType);
        return creator.create(viewGroup);
    }

    @Override
    public void onBindViewHolder(@NonNull BaseViewHolder holder, int position) {
        holder.bind(mContext, (MultiTypeEntity) getItem(position), null, position, mObjects);
    }

    @Override
    public void onBindViewHolder(@NonNull BaseViewHolder holder, int position, @NonNull List<Object> payloads) {
        super.onBindViewHolder(holder, position, payloads);
        holder.bind(mContext, (MultiTypeEntity) getItem(position), payloads, position, mObjects);
    }

    public Object getItem(int position) {
        return mDataList != null ? mDataList.get(position) : null;
    }

    @Override
    public int getItemCount() {
        return mDataList != null ? mDataList.size() : 0;
    }

    @Override
    public int getItemViewType(int position) {
        MultiTypeEntity entity = (MultiTypeEntity) getItem(position);
        return entity.getItemType();
    }
}

File: library/src/main/java/com/lzx/library/EfficientAdapter.kt
package com.lzx.library

import android.support.v4.util.SparseArrayCompat
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup

open class EfficientAdapter<T> : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private const val FALLBACK_DELEGATE_VIEW_TYPE = Int.MAX_VALUE - 1
    }

    var items: MutableList<T>? = mutableListOf()
    private val typeHolders: SparseArrayCompat<ViewHolderCreator<T>> = SparseArrayCompat()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val holder = getHolderForViewType(viewType)
                ?: throw NullPointerException("No Holder added for ViewType $viewType")
        return BaseViewHolder(parent, holder.getResourceId())
    }

    override fun getItemCount(): Int = items?.size ?: 0

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        onBindViewHolder(viewHolder, position, mutableListOf())
    }

    override fun onBindViewHolder(
            viewHolder: RecyclerView.ViewHolder, position: Int, payloads: MutableList<Any>
    ) {
        val holder = getHolderForViewType(viewHolder.itemViewType)
                ?: throw java.lang.NullPointerException(
                        "No Holder added for ViewType $viewHolder.itemViewType")
        holder.registerItemView(viewHolder.itemView)
        holder.onBindViewHolder(items?.get(position), items, position, holder, payloads)
    }

    override fun getItemViewType(position: Int): Int {
        if (items == null) {
            throw NullPointerException("adapter data source is null")
        }
        for (i in 0 until typeHolders.size()) {
            val holder = typeHolders.valueAt(i)
            val data = items!!.getOrNull(position)
            if (holder.isForViewType(data, position)) {
                return typeHolders.keyAt(i)
            }
        }
        val errorMessage: String
        errorMessage = if (items is List<*>) {
            val itemString = (items as List<*>)[position].toString()
            "No holder added that matches item=$itemString at position=$position in data source"
        } else {
            "No holder added for item at position=$position. items=$items"
        }
        throw java.lang.NullPointerException(errorMessage)
    }

    private fun getHolderForViewType(viewType: Int): ViewHolderCreator<T>? {
        return typeHolders.get(viewType)
    }

    fun addTypeHolder(holder: ViewHolderCreator<T>): EfficientAdapter<T> {
        var viewType: Int = typeHolders.size()
        while (typeHolders.get(viewType) != null) {
            viewType++
            require(
                    viewType != FALLBACK_DELEGATE_VIEW_TYPE) {
                "Oops, we are very close to Integer.MAX_VALUE. " +
                        "It seems that there are no more free and " +
                        "unused view type integers left to add another holder."
            }
        }
        return addTypeHolder(viewType, holder)
    }

    fun addTypeHolder(
            viewType: Int,
            holder: ViewHolderCreator<T>?
    ): EfficientAdapter<T> {
        if (holder == null) {
            throw java.lang.NullPointerException("holder is null!")
        }
        require(viewType != FALLBACK_DELEGATE_VIEW_TYPE) {
            ("The view type = " + FALLBACK_DELEGATE_VIEW_TYPE +
                    " is reserved for fallback adapter holder Please use another view type.")
        }
        require(typeHolders.get(viewType) == null) {
            ("An holder is already registered for the viewType = " +
                    viewType +
                    ". Already registered holder is " + typeHolders.get(viewType))
        }
        typeHolders.put(viewType, holder)
        return this
    }

    open fun insertedData(position: Int, data: T) = apply {
        items?.add(position, data)
        notifyItemInserted(position)
    }

    open fun removedData(position: Int) = apply {
        items?.removeAt(position)
        notifyItemRemoved(position)
    }

    open fun updateData(position: Int, data: T, payload: Boolean = true) {
        items?.set(position, data)
        if (payload) {
            notifyItemChanged(position, data)
        } else {
            notifyItemChanged(position)
        }
    }

    fun getItem(position: Int): T? = items?.getOrNull(position)

    fun register(holder: ViewHolderCreator<T>): EfficientAdapter<T> {
        var viewType: Int = typeHolders.size()
        while (typeHolders.get(viewType) != null) {
            viewType++
            require(
                    viewType != FALLBACK_DELEGATE_VIEW_TYPE) {
                "Oops, we are very close to Integer.MAX_VALUE. " +
                        "It seems that there are no more free" +
                        "and unused view type integers left to add another holder."
            }
        }
        return register(viewType, holder)
    }

    fun register(viewType: Int, holder: ViewHolderCreator<T>): EfficientAdapter<T> {
        require(viewType != FALLBACK_DELEGATE_VIEW_TYPE) {
            ("The view type = $FALLBACK_DELEGATE_VIEW_TYPE is reserved " +
                    "for fallback adapter holder). Please use another view type.")
        }
        typeHolders.put(viewType, holder)
        return this
    }

    fun attach(recyclerView: RecyclerView): EfficientAdapter<T> {
        recyclerView.adapter = this
        return this
    }

    fun submitList(list: MutableList<T>) {
        this.items?.clear()
        this.items?.addAll(list)
        notifyDataSetChanged()
    }
}